numpy/_core/src/multiarray/scalartypes.cpp

```diff
@@ -3170,35 +3835,84 @@ gentype_arrtype_getbuffer(PyObject *self, Py_buffer *view, int flags)
     return 0;
 }
 
+template <typename PT>
+constexpr static const char *
+get_struct_fmt_for_npy_type()
+{
+    // cannot use npy_<type> because they might be
+    // the same underlying type but with different name
+    // and should be treated differently
+    // e.g.
+    // static_assert(std::is_same_v<npy_ubyte, npy_bool>);
+    // static_assert(std::is_same_v<npy_ushort, npy_half>);
+    // static_assert(std::is_same_v<npy_double, npy_longdouble>);
 
-static PyBufferProcs gentype_arrtype_as_buffer = {
-    .bf_getbuffer = (getbufferproc)gentype_arrtype_getbuffer,
-};
-
-
-/**begin repeat
- * #name = bool, byte, short, int, long, longlong, ubyte, ushort, uint, ulong,
- *         ulonglong, half, float, double, longdouble, cfloat, cdouble,
- *         clongdouble#
- * #Name = Bool, Byte, Short, Int, Long, LongLong, UByte, UShort, UInt, ULong,
- *         ULongLong, Half, Float, Double, LongDouble, CFloat, CDouble,
- *         CLongDouble#
- * #NAME = BOOL, BYTE, SHORT, INT, LONG, LONGLONG, UBYTE, USHORT, UINT, ULONG,
- *         ULONGLONG, HALF, FLOAT, DOUBLE, LONGDOUBLE, CFLOAT, CDOUBLE,
- *         CLONGDOUBLE#
- * #fmt = ?, b, h, i, l, q, B, H, I, L, Q, e, f, d, g, Zf, Zd, Zg#
- */
+    if constexpr (std::is_same_v<PT, PyBoolScalarObject>) {
+        return "?";
+    }
+    else if constexpr (std::is_same_v<PT, PyByteScalarObject>) {
+        return "b";
+    }
+    else if constexpr (std::is_same_v<PT, PyShortScalarObject>) {
+        return "h";
+    }
+    else if constexpr (std::is_same_v<PT, PyIntScalarObject>) {
+        return "i";
+    }
+    else if constexpr (std::is_same_v<PT, PyLongScalarObject>) {
+        return "l";
+    }
+    else if constexpr (std::is_same_v<PT, PyLongLongScalarObject>) {
+        return "q";
+    }
+    else if constexpr (std::is_same_v<PT, PyUByteScalarObject>) {
+        return "B";
+    }
+    else if constexpr (std::is_same_v<PT, PyUShortScalarObject>) {
+        return "H";
+    }
+    else if constexpr (std::is_same_v<PT, PyUIntScalarObject>) {
+        return "I";
+    }
+    else if constexpr (std::is_same_v<PT, PyULongScalarObject>) {
+        return "L";
+    }
+    else if constexpr (std::is_same_v<PT, PyULongLongScalarObject>) {
+        return "Q";
+    }
+    else if constexpr (std::is_same_v<PT, PyHalfScalarObject>) {
+        return "e";
+    }
+    else if constexpr (std::is_same_v<PT, PyFloatScalarObject>) {
+        return "f";
+    }
+    else if constexpr (std::is_same_v<PT, PyDoubleScalarObject>) {
+        return "d";
+    }
+    else if constexpr (std::is_same_v<PT, PyLongDoubleScalarObject>) {
+        return "g";
+    }
+    else if constexpr (std::is_same_v<PT, PyCFloatScalarObject>) {
+        return "Zf";
+    }
+    else if constexpr (std::is_same_v<PT, PyCDoubleScalarObject>) {
+        return "Zd";
+    }
+    else if constexpr (std::is_same_v<PT, PyCLongDoubleScalarObject>) {
```

Comment:
general question about code like this from a non-expert in C++, code like this using `constexpr` is evaluated at compile-time, right?